rust/ql/test/library-tests/type-inference/main.rs

```diff
@@ -1810,6 +1810,104 @@ mod indexers {
     }
 }
 
+mod loops {
+    struct MyCallable {
+    }
+
+    impl MyCallable {
+        fn new() -> Self {
+            MyCallable {}
+        }
+
+        fn call(&self) -> i64 {
+            1
+        }
+    }
+
+    pub fn f() {
+        // for loops with arrays
+
+        for i in [1, 2, 3] { } // $ type=i:i32
+        for i in [1, 2, 3].map(|x| x + 1) { } // $ MISSING: type=i:i32
+        for i in [1, 2, 3].into_iter() { } // $ MISSING: type=i:i32
+
+        let vals1 = [1u8, 2, 3]; // $ MISSING: type=vals1:[u8; 3]
+        for u in vals1 { } // $ type=u:u8
+
+        let vals2 = [1u16; 3]; // $ MISSING: type=vals2:[u16; 3]
+        for u in vals2 { } // $ type=u:u16
+
+        let vals3: [u32; 3] = [1, 2, 3]; // $ MISSING: type=vals3:[u32; 3]
+        for u in vals3 { } // $ type=u:u32
+
+        let vals4: [u64; 3] = [1; 3]; // $ MISSING: type=vals4:[u64; 3]
+        for u in vals4 { } // $ type=u:u64
+
+        let mut strings1 = ["foo", "bar", "baz"]; // $ MISSING: type=strings1:[&str; 3]
+        for s in &strings1 { } // $ MISSING: type=s:&str
+        for s in &mut strings1 { } // $ MISSING: type=s:&str
+        for s in strings1 { } // $ type=s:str
+
+        let strings2 = [String::from("foo"), String::from("bar"), String::from("baz")]; // $ MISSING: type=strings2:[String; 3]
+        for s in strings2 { } // $ type=s:String
+
+        let strings3 = &[String::from("foo"), String::from("bar"), String::from("baz")]; // $ MISSING: type=strings3:&[String; 3]
+        for s in strings3 { } // $ MISSING: type=s:String
+
+        let callables = [MyCallable::new(), MyCallable::new(), MyCallable::new()]; // $ MISSING: type=callables:[MyCallable; 3]
+        for c in callables { // $ type=c:MyCallable
+            let result = c.call(); // $ type=result:i64 method=call
+        }
+
+        // for loops with ranges
+
+        for i in 0..10 { } // $ MISSING: type=i:i32
+        for u in [0u8 .. 10] { } // $ MISSING: type=u:u8
+
+        let range1 = std::ops::Range { start: 0u16, end: 10u16 }; // $ MISSING: type=range:std::ops::Range<u16>
+        for u in range1 { } // $ MISSING: type=i:u16
+
+        // for loops with containers
+
+        let vals3 = vec![1, 2, 3]; // $ MISSING: type=vals3:Vec<i32>
```

Comment:
This is not how to make expected type annotations for constructed types, instead you should use the format `type=<element>:<path>.<type>` (see https://github.com/github/codeql/blob/2e5d52e7e8eb101a9e209dd9c232205045a3b58b/rust/ql/test/library-tests/type-inference/type-inference.ql#L60). So in this case it should instead by `type=vals3:T.i32`.

Reply:
Ah, I think I was relying on the test runner to start suggesting correct syntax, which it doesn't really do for optional results.  I've updated the expectations now as best I can - there may still be add mistakes.

Reply:
This one still needs to be updated.